### ISSUE TITLE
Switch to Github Actions

### ISSUE THREAD

**Issue by lionel-** [emacs-ess/ESS#1079]
I see this on travis-CI:

```
Please be aware travis-ci.org will be shutting down in several weeks, with all accounts migrating to travis-ci.com. Please stay tuned here for more information. 
```

If possible, we should take this opportunity to add checks on a Windows platform.

**Comment by jabranham** [comment]
I'm a little hesitant to switch to something that deepens our dependence on github/microsoft. I think gitlab CI runs on open sourced code, perhaps we should take a look at that?

**Comment by lionel-** [comment]
I trust Microsoft to keep properly funding the development and computation power of the github CI for the next 15 to 20 years. I can't say as much about gitlab.

github has been working well for us. It would take a lot of investigation, time, and energy to switch to gitlab, which we don't have.

Finally, it'd reassure me to set up a Windows CI before the next release. Github actions seems the fastest way to get there.

**Comment by lionel-** [comment]
On a more positive note I think it might be worthwhile to investigate alternatives but I'd much prefer that to happen after the next release. Until then we need a quick solution for the CI so we can finish the release.

**Comment by mmaechler** [comment]
> I'm a little hesitant to switch to something that deepens our dependence on github/microsoft. I think gitlab CI runs on open sourced code, perhaps we should take a look at that?

Thank you, @jabranham;  entirely my feelings, too.   More generally, be aware that the extreme CI (every commit .. unless you do a special hoop jump) is also adding to the huge energy consumption recent IT has brought, and is hence contributing to climate heating.. 

Honestly, I don't see why CI is necessary for finishing the release.   Our `make` based checking should work on our own computers, shouldn't it ?
Last but not least, I'm very grateful to @lionel-  for emphasizing to work towards release  (and hence refrain from new features, etc.. which are not helping release).

**Comment by lionel-** [comment]
The CI happens on pushes (which can include several commits) and pull requests (whole branches), not necessarily on every commit. Unfortunately, checking locally is not sufficient because we need to check on all supported versions of Emacs and also on Windows. We currently don't test on Windows but I would like to do it before the next release. We encountered Windows-only freezes in #1043, and Windows-only timeouts in aliases retrievals (#1025 and #1081).

**Comment by lionel-** [comment]
I'll add that tests for interaction with the REPL are hard to set up. I regularly get a failing CI even though the checks pass locally. This happened again today in #1082 where my tests were missing some wait-for-process calls. ESS is complex and sits on top of a cross-platform IPC stack. This is why it needs a good test suite and continuous checking on different platforms.

**Comment by lionel-** [comment]
Relevant twitter post about the support on Windows: https://twitter.com/NoorahAlhasan/status/1344008026290860033

**Comment by mmaechler** [comment]
Could we consider adopting the same / a similar  __CI only opt-in__  scheme that @bbolker has been proposing and implementing for `lme4` (see https://github.com/lme4/lme4/commit/244d65d830093fe52ebe85fb1ffdc8ee7d63522b .. but maybe ask Ben)

**Comment by lionel-** [comment]
Yes, though I'd prefer not cluttering the git history with CI instructions. Another way to do this is to enable CI only on a special branch on which we would push to get a CI run.

However, I think we should keep the CI as is:

1. Our CI load is very modest (runs in 2 minutes). We probably use much more energy by developing dev ESS and distributing it then testing it on CI.

2. Our development branch is immediately put into "production" via MELPA. It is very useful to have the dev branch so widely tested (lots of bug reports) and we owe it to our dev users to ensure that the development version always passes CI.

**Comment by lionel-** [comment]
We could limit our workload by testing:

- On the oldest supported Emacs version (macOS)
- On the current Emacs version (Windows)
- On dev Emacs (Linux)

Currently we are systematically testing all versions between oldest and dev which make a total of 6 builds.

**Comment by vspinu** [comment]
Maybe a better use of effort in the mid-short term would be to have an automated docker-fu to run tests on Windows locally. Also running interactively with virtenv would be nice. Also to interactively test remotes, one can use dockers. 

I am personally not against github CI. It's nice, but feels a bit raw still. Configuration and maintenance is still a bit tricky I think. Moving the entire infra for the sake of Win tests feels a bit unwarranted. 

We can try having only one Win build first, and see how it goes (in case someone has time and wants to attempt the exercise).

**Comment by lionel-** [comment]
oh I missed that: https://blog.travis-ci.com/oss-announcement

Apparently travis will remain free for open source. We just need to migrate from travis-ci.org to travis-ci.com.
